### PR TITLE
Colorized logging

### PR DESCRIPTION
**Pull request by gurupras**
Short Description: Updates to the (updated) logging system

Fixes:
- Adds lost colour to the event system
- Adds new methods to loggers

Some of the colours may not be the same as they were in the earlier `logger.log` system. I can change these if you're very particular. Some of the LCD code has also crept in. I can get rid of these if they're no longer required.

The commit in itself may have some redundant code in terms of functions that can be used like:
`self.logger.red('message in red')`, `...`
I thought these might be useful for cases where a message doesn't necessarily fit into the usual logging levels like `CRITICAL`, `ERROR`, `WARNING`, `INFO`, `DEBUG`

### CODE REVIEW

**Comment by douglascamata** [comment]
@gurupras please, remember that this colors cannot be logged in files (and there will be options for the user to log every task, or a specific task, to a file), so we need to handle this.

**Comment by gurupras** [comment]
I see. I may have missed this, but all the code I found was only invoking `logging.basicConfig` without passing in a file path. Regardless, if I understood the design correctly, this is trivial to support if it involves a new handler akin to `socketio_handler.py` and `logging_handler.py`

**Comment by gurupras** [comment]
Oh, my bad. You're talking about `getattr(logger, level)(message)`. We can still handle this while still providing colorized output on the console as long as the color formatter is not attached to the file handler.

**Comment by MrCarlosJr** [comment]
input option -V[erbose] -c[olor]

**Comment by raulgbcr** [comment]
This NEEDS to be implemented, actual gui is shit.

**Comment by gurupras** [comment]
Well, not necessarily.
Actual GUI might be nice to have..especially if we can incorporate a map
and stuff.

However, you're right. I would still like colorized logging available
regardless of GUI.
You just *immediately* know to focus on what's important and ignore what's
not.

Let me know if this requires any additional work, I'll be happy to get it
done.

**Comment by douglascamata** [comment]
@raulgbcr if by actual GUI you meant the terminal, it's NOT supposed to be user friendly from the start. It's not a GUI, it's an application with logs. We have no GUI (:

**Comment by douglascamata** [comment]
👍 

@TheSavior please review this

File: pokemongo_bot/event_manager.py [hunk 1]
```diff
@@ -26,9 +26,10 @@ def __init__(self, *handlers):
         self._handlers = handlers or []
 
     def event_report(self):
-        for event, parameters in self._registered_events.iteritems():
+        for event, (color, parameters) in self._registered_events.iteritems():
             print '-'*80
             print 'Event: {}'.format(event)
+            print 'Color: {}'.format(color)
```
Comment (douglascamata) [inline pokemongo_bot/event_manager.py]: please, remove these prints.

File: pokemongo_bot/logger.py [hunk 2]
```diff
@@ -18,3 +28,112 @@ def log(msg, color=None):
...
+root = logging.getLogger()
+if lcd:
+    root.addHandler(logger.LcdHandler())
+
+def init_logger():
+    rootLogger = logging.getLogger()
+    for handler in rootLogger.handlers:
+        handler.setFormatter(LogFormatter())
+
+
```
Comment (douglascamata) [inline pokemongo_bot/logger.py]: Please, removing anything not related to your PR from this file, like the LcdHandler.

File: pokemongo_bot/event_manager.py [hunk 3]
```diff
@@ -29,7 +29,6 @@ def event_report(self):
         for event, (color, parameters) in self._registered_events.iteritems():
             print '-'*80
             print 'Event: {}'.format(event)
```
Comment (douglascamata) [inline pokemongo_bot/event_manager.py]: There is still this print and the print above.

Comment (gurupras) [inline pokemongo_bot/event_manager.py]: I don't think those were added by me. But sure, I'll remove them.

Comment (douglascamata) [inline pokemongo_bot/event_manager.py]: oh shit, my bad... don't remove those, here it's the event's report... they should be here

**Comment by douglascamata** [comment]
@gurupras please put the event name print back, I forgot that code was inside the event report function. It should indeed be there. My bad.

**Comment by gurupras** [comment]
But you don't want the port print? or you want both back

**Comment by gurupras** [comment]
oops..sorry not port..it was:

            print '-'*80
            print 'Event: {}'.format(event)

File: pokemongo_bot/logger.py [hunk 4]
```diff
@@ -18,3 +17,106 @@ def log(msg, color=None):
     logger = logging.getLogger('generic')
     logger.info(msg)
     warnings.warn(message, DeprecationWarning)
+
+class LcdHandler(logging.Handler):
```
Comment (douglascamata) [inline pokemongo_bot/logger.py]: The LcdHandler is beyond the objective of this PR and being reimplemented by @brantje. Can you remove it, please?

**Comment by raulgbcr** [comment]
@douglascamata "Terminal is not suposed to be user friendly" ?

Then whats suposed to be? Lot of people prefer the colors for something, like @gurupras said we can know important things and things we dont need to care about. 

Btw, check out the file for the `map teleport` feature, it is not updated to the event system.

**Comment by douglascamata** [comment]
@raulgbcr terminal is supposed to be developer friendly that presents logs, not a story that even a 9 years old can read. Adding colors can be nice, but it's just a plus, not a requirement. 

I know about the map teleport. I created a fallback because ppl were adding more logs than I could fix and it was taking a very long time to merge the event system. There's a PR fixing it already.

**Comment by raulgbcr** [comment]
@gurupras Perfect, thank you! ^^

**Comment by douglascamata** [comment]
👍 

@TheSavior, please review this.

**Comment by TheSavior** [comment]
I don't think I have enough context here to provide meaningful feedback, so here are some general thoughts on a logging system with color:

We emit events of a given type with some data.
Different things can listen to those, GUI apps, console, files, etc.
I'd imagine for the console part, there would be a thing that would listen to the triggered events and write them to the console.
I would expect that class would be the thing that would contain the mapping from event type to color to be able to write them to the console.
This way other listeners (file, gui, etc), don't need to have any concept of colors for a given event.

**Comment by gurupras** [comment]
This (I think) is how it is structured currently. We currently have two handlers `websocket_handler.py` and `logging_handler.py` which are designed to do two different things. 

All this patch creates is a new `Formatter` in the logging framework. Although currently this `Formatter` is applied to all handlers of the logging framework, it can certainly be changed to apply to only the console handler.

**Comment by gurupras** [comment]
Sorry, that last comment a was little confusing.

Okay, so currently, we have two `event_handlers`. These are: `logging_handler` and `socketio_handler`. Only the `logging_handler` calls `logger.colorized(...)` which produces colorized output.  While this may sound reasonable, it is not entirely correct. This is because, the `logging_handler` is probably supposed to handle logging to file and/or console. Currently, both of these would be affected by this PR (which we don't want). 

When we do finally have file logging changes in place (do we have it already?), we need to make a small tweak to ensure that the `Formatter` that produces colorized output is only applied to the `Handler` that handles the console. This will ensure that the `Handler` handling files will not produce any of the weird ANSI color codes.

**Comment by TheSavior** [comment]
Hmm, I guess the part that I saw that made me unsure how things were implemented was seeing this change in pokecli:

```
 self.event_manager.register_event('api_error')
```

becoming this:

```
self.event_manager.register_event('api_error', color='red')
```

It seems like we should be setting the color wherever we are listening to the events and writing them to the console instead of telling every handler what color to use and having them ignore it.

**Comment by gurupras** [comment]
That's a good point. I'm not too familiar with the whole event system. I don't see any easy way to achieve this. The reason is because the `logging_handler` has no notion of what events are available and is not informed of registrations. It is written to be simple and if this handler is registered, it just blindly logs the events. 

Perhaps we should wait for the people who created the event system to provide comments.

**Comment by gurupras** [comment]
> It seems like we should be setting the color wherever we are listening to the events and writing them to the console instead of telling every handler what color to use and having them ignore it.

While this sounds like the right thing to do, I think it is very difficult to implement correctly. This is because you either a) want handlers to know about what events exist in the system or b) manually inform handler(s) of events and their associated colors.

**Comment by douglascamata** [comment]
@TheSavior sometimes just adding more parameters to `event_manager.emit` doesn't seem right, because it would imply in big code repetition. 

@gurupras I like your solution and I'm the event system's creator :P

**Comment by gurupras** [comment]
This is probably not the right place to ask this, I will take this to Slack if required, but I feel like, on some level, the event system is re-doing a lot of the same work that already exists in the `logging` framework. I'm sure we can do socketio emits inside a `logging.Handler` and treat the `record` as handled.

**Comment by TheSavior** [comment]
@douglascamata What if the web ui wants to start doing special things for given events. Should those be passed in the initial data too and thus sent to all listeners? The idea is that you'd be keeping handler specific functionality inside the handler instead of sending unnecessary data around.

**Comment by abusalman** [comment]
While testing this PR, I got this error:
```
11:12:56 [PokemonGoBot] [Level info] [position_update] Now at (40.778866, -73.969417, 0)
11:12:56 [PokemonGoBot] [Level debug] [load_cached_location] Loading cached location...
11:12:56 [PokemonGoBot] [Level debug] [location_cache_ignored] Going to a new place, ignoring cached location.
11:12:56 [PokemonGoBot] [Level info] [login_started] Login procedure started.
Traceback (most recent call last):
  File "pokecli.py", line 497, in <module>
    main()
  File "pokecli.py", line 131, in main
    report_summary(bot)
  File "pokecli.py", line 140, in report_summary
    metrics.capture_stats()
  File "c:\pokebot\PokemonGo-Bot\pokemongo_bot\metrics.py", line 73, in capture_stats
    request = self.bot.api.create_request()
  File "c:\pokebot\PokemonGo-Bot\pokemongo_bot\api_wrapper.py", line 28, in create_request
    self._position_alt
  File "c:\pokebot\PokemonGo-Bot\pokemongo_bot\api_wrapper.py", line 44, in __init__
    PGoApiRequest.__init__(self, *args)
TypeError: __init__() takes exactly 5 arguments (6 given)
11:12:59 [sentry.errors] [ERROR] Sentry responded with an error: 'ascii' codec can't decode byte 0x9c in position 1: ordinal not in range(128) (url: https://app.getsentry.com/api/90254/store/)
Traceback (most recent call last):
  File "C:\pokebot\PokemonGo-Bot\lib\site-packages\raven\transport\threaded.py", line 174, in send_sync
    super(ThreadedHTTPTransport, self).send(data, headers)
  File "C:\pokebot\PokemonGo-Bot\lib\site-packages\raven\transport\http.py", line 47, in send
    ca_certs=self.ca_certs,
  File "C:\pokebot\PokemonGo-Bot\lib\site-packages\raven\utils\http.py", line 66, in urlopen
    return opener.open(url, data, timeout)
  File "C:\pokebot\PokemonGo-Bot\lib\site-packages\future\backports\urllib\request.py", line 494, in open
    response = self._open(req, data)
  File "C:\pokebot\PokemonGo-Bot\lib\site-packages\future\backports\urllib\request.py", line 512, in _open
    '_open', req)
  File "C:\pokebot\PokemonGo-Bot\lib\site-packages\future\backports\urllib\request.py", line 466, in _call_chain
    result = func(*args)
  File "C:\pokebot\PokemonGo-Bot\lib\site-packages\raven\utils\http.py", line 46, in https_open
    return self.do_open(ValidHTTPSConnection, req)
  File "C:\pokebot\PokemonGo-Bot\lib\site-packages\future\backports\urllib\request.py", line 1284, in do_open
    h.request(req.get_method(), req.selector, req.data, headers)
  File "c:\python27\Lib\httplib.py", line 1053, in request
    self._send_request(method, url, body, headers)
  File "c:\python27\Lib\httplib.py", line 1093, in _send_request
    self.endheaders(body)
  File "c:\python27\Lib\httplib.py", line 1049, in endheaders
    self._send_output(message_body)
  File "c:\python27\Lib\httplib.py", line 891, in _send_output
    msg += message_body
UnicodeDecodeError: 'ascii' codec can't decode byte 0x9c in position 1: ordinal not in range(128)
11:12:59 [sentry.errors.uncaught] [ERROR] [u'TypeError: __init__() takes exactly 5 arguments (6 given)', u'  File "pokecli.py", line 497, in <module>', u'  File "pokecli.py", line 131, in main', u'  File "pokecli.py", line 140, in report_summary', u'  File "c:\\pokebot\\PokemonGo-Bot\\pokemongo_bot\\metrics.py", line 73, in capture_stats', u'  File "c:\\pokebot\\PokemonGo-Bot\\pokemongo_bot\\api_wrapper.py", line 28, in create_request', u'  File "c:\\pokebot\\PokemonGo-Bot\\pokemongo_bot\\api_wrapper.py", line 44, in __init__']

```

**Comment by BartKoppelmans** [comment]
How about using different types of logging functions (e.g. error, pokémon, item) to allow for more extensibility? If we would use one logging class, we could give it multiple functions to call. Then the function could use its own color and settings (possibly shared across the class). This would make it easier to add options in the future, for example to allow setting the color or to disable certain logging. 

A con is that we would have to specify a type in the code.

@douglascamata: How much work would it be to have the events use different functions, according to a given event? And what do you thing about that kind of system?

**Comment by TheSavior** [comment]
@BartKoppelmans When you receive an event, you know what type of event it is. Anything you could do with multiple functions you could do inside of the event handler.

**Comment by BartKoppelmans** [comment]
@TheSavior: Nice, so it would be possible to implement this? What do you think about it then?

**Comment by douglascamata** [comment]
> @douglascamata: How much work would it be to have the events use different functions, according to a given event? And what do you thing about that kind of system?

We want to avoid duplicating code and be as simple as possible for the requirements we have. Why would we need multiple functions? What is so special that cannot be handled dynamically in just one function?

**Comment by douglascamata** [comment]
> @douglascamata What if the web ui wants to start doing special things for given events. Should those be passed in the initial data too and thus sent to all listeners? The idea is that you'd be keeping handler specific functionality inside the handler instead of sending unnecessary data around.

That is not in our requirements, at least yet. So don't need to bother thinking about it because we don't know if that will ever happen (and personally, I don't think it will).

**Comment by BartKoppelmans** [comment]
@douglascamata 
> We want to avoid duplicating code and be as simple as possible for the requirements we have. Why would we need multiple functions? What is so special that cannot be handled dynamically in just one function?

This would enable us to have different functions for each type of log, and thus allow us to have separate settings for these functions. We could say no to logging the incubator and yes to logging the rest, while keeping the caught pokémon green and the transfers red. This would make it easier to extend, without having to change the old code. The common code can still be written in one function, which is called to reduce redundancy. 

The biggest plus is the extensibility, as it would make it easier to add settings to the functions. Doing this dynamically would result in one method full of complex nesting and if-statements.

**Comment by gurupras** [comment]
> The biggest plus is the extensibility, as it would make it easier to add settings to the functions. Doing this dynamically would result in one method full of complex nesting and if-statements.

Or a json array of events that we want to log

**Comment by douglascamata** [comment]
@BartKoppelmans @gurupras I already have plans for this and it will keep everything as dynamic as it is right now. There will be configs in every task inheriting from `BaseTask` to customize log level, format and destination (file or terminal).

**Comment by BartKoppelmans** [comment]
@douglascamata I'm curious, share it as soon as you start! 😄

**Comment by douglascamata** [comment]
@BartKoppelmans I will when I do it, but i'm working on more important stuff and I'm a bit overloaded.

**Comment by douglascamata** [comment]
@BartKoppelmans btw, can you solve merge conflicts, please?

**Comment by BartKoppelmans** [comment]
@douglascamata Sure, let me know if I can help with some things! And I'm sorry, but this is not my PR, otherwise I would!😅 

Pinging @gurupras to resolve the merge conflicts.

**Comment by BartKoppelmans** [comment]
Great job, thank you very much! @gurupras

**Comment by DanielVolz** [comment]
can someone help me, getting this errors on OSX with iterm app.

`Traceback (most recent call last):
  File "/usr/local/Cellar/python/2.7.12/Frameworks/Python.framework/Versions/2.7/lib/python2.7/logging/__init__.py", line 861, in emit
    msg = self.format(record)
  File "/usr/local/Cellar/python/2.7.12/Frameworks/Python.framework/Versions/2.7/lib/python2.7/logging/__init__.py", line 734, in format
    return fmt.format(record)
  File "/Users/danielvolz/MyPogoHeroku/PokemonGo-Bot/pokemongo_bot/logger.py", line 55, in format
    record.msg = '{color}{message}{reset}'.format(color=ColorizedLogFormatter.COLOR_SEQ % (color), message=record.msg, reset=ColorizedLogFormatter.RESET_SEQ)
UnicodeEncodeError: 'ascii' codec can't encode character u'\xf6' in position 48: ordinal not in range(128)
Logged from file logger.py, line 86
22:59:13 [RecycleItems] [Level info] [item_discarded] Discarded 1x Pokeball (maximum 100).
22:59:14 [RecycleItems] [Level info] [item_discarded] Discarded 1x Super Potion (maximum 0).
Traceback (most recent call last):
  File "/usr/local/Cellar/python/2.7.12/Frameworks/Python.framework/Versions/2.7/lib/python2.7/logging/__init__.py", line 861, in emit
    msg = self.format(record)
  File "/usr/local/Cellar/python/2.7.12/Frameworks/Python.framework/Versions/2.7/lib/python2.7/logging/__init__.py", line 734, in format
    return fmt.format(record)
  File "/Users/danielvolz/MyPogoHeroku/PokemonGo-Bot/pokemongo_bot/logger.py", line 55, in format
    record.msg = '{color}{message}{reset}'.format(color=ColorizedLogFormatter.COLOR_SEQ % (color), message=record.msg, reset=ColorizedLogFormatter.RESET_SEQ)
UnicodeEncodeError: 'ascii' codec can't encode character u'\xf6' in position 48: ordinal not in range(128)
Logged from file logger.py, line 86
Traceback (most recent call last):
  File "/usr/local/Cellar/python/2.7.12/Frameworks/Python.framework/Versions/2.7/lib/python2.7/logging/__init__.py", line 861, in emit
    msg = self.format(record)
  File "/usr/local/Cellar/python/2.7.12/Frameworks/Python.framework/Versions/2.7/lib/python2.7/logging/__init__.py", line 734, in format
    return fmt.format(record)
  File "/Users/danielvolz/MyPogoHeroku/PokemonGo-Bot/pokemongo_bot/logger.py", line 55, in format
    record.msg = '{color}{message}{reset}'.format(color=ColorizedLogFormatter.COLOR_SEQ % (color), message=record.msg, reset=ColorizedLogFormatter.RESET_SEQ)
UnicodeEncodeError: 'ascii' codec can't encode character u'\xf6' in position 48: ordinal not in range(128)
Logged from file logger.py, line 86
Traceback (most recent call last):
  File "/usr/local/Cellar/python/2.7.12/Frameworks/Python.framework/Versions/2.7/lib/python2.7/logging/__init__.py", line 861, in emit
    msg = self.format(record)
  File "/usr/local/Cellar/python/2.7.12/Frameworks/Python.framework/Versions/2.7/lib/python2.7/logging/__init__.py", line 734, in format
    return fmt.format(record)
  File "/Users/danielvolz/MyPogoHeroku/PokemonGo-Bot/pokemongo_bot/logger.py", line 55, in format
    record.msg = '{color}{message}{reset}'.format(color=ColorizedLogFormatter.COLOR_SEQ % (color), message=record.msg, reset=ColorizedLogFormatter.RESET_SEQ)
UnicodeEncodeError: 'ascii' codec can't encode character u'\xf6' in position 48: ordinal not in range(128)
Logged from file logger.py, line 86
Traceback (most recent call last):
  File "/usr/local/Cellar/python/2.7.12/Frameworks/Python.framework/Versions/2.7/lib/python2.7/logging/__init__.py", line 861, in emit
    msg = self.format(record)
  File "/usr/local/Cellar/python/2.7.12/Frameworks/Python.framework/Versions/2.7/lib/python2.7/logging/__init__.py", line 734, in format
    return fmt.format(record)
  File "/Users/danielvolz/MyPogoHeroku/PokemonGo-Bot/pokemongo_bot/logger.py", line 55, in format
    record.msg = '{color}{message}{reset}'.format(color=ColorizedLogFormatter.COLOR_SEQ % (color), message=record.msg, reset=ColorizedLogFormatter.RESET_SEQ)
UnicodeEncodeError: 'ascii' codec can't encode character u'\xf6' in position 48: ordinal not in range(128)
Logged from file logger.py, line 86`

**Comment by gurupras** [comment]
Oh boy. 
I don't have access to a Mac. Could someone who uses OSX take a look at this and suggest a fixup?

**Comment by douglascamata** [comment]
@gurupras logger.py needs to import unicode_literals from future, probably.

**Comment by DanielVolz** [comment]
trying now with the new commit... looks good right now

**Comment by douglascamata** [comment]
@gurupras I'm sorry, but #3198 had a much more modular structure and it was merged :(

**Pull request closed without merging**